### PR TITLE
[api] support staging without write access in GA projects

### PR DESCRIPTION
**Pull request by adrianschroeter**
SLSA requires seperation of release managers and code stream managers.
Therefore release managers just approve reviews and the code stream
managers using the request approve mechanism.

As consequence the release managers have no write access anymore
in the GA projects:
 * request gets not accepted directly via staging api, just indirect
   via approve mechanism
 * modifications outside of the requests (like fiddeling with local links)
   are not allowed anymore

jsc#OBS-200

### CODE REVIEW

**Comment by adrianschroeter** [comment]
this is just currently an untested draft, needs further investigation and testing. And esp. review of permission handling.

**Comment by eduardoj** [comment]
@adrianschroeter, only a small note about labelling pull requests: we usually create issues and label them with a priority label (P2 for example). And after that we create pull requests which could solve the issue. This helps us to separate the problem (and assign to it a prioriy) from the solution.

File: src/api/app/models/project.rb
Comment (coolo) [inline src/api/app/models/project.rb:1127]: And so this call will find requests with by_project review and approver - but you're looking for unapproved requests (according to the name of the function) and those do *not* have an approver. 

What you really want to look at is if all staged requests are approved - those are listed in `@staging_project.staged_requests`.

File: src/api/app/controllers/staging/staging_projects_controller.rb
Comment (coolo) [inline src/api/app/controllers/staging/staging_projects_controller.rb:51]: That should be part of staging project accept policy instead. Skipping authorize calls like this makes them basically untestable as policies - and the whole claim behind not testing controller permissions is that policies in itself are tested (IMO)

**Review by coolo — CHANGES_REQUESTED** [review]

File: src/api/app/models/project.rb
Comment (hennevogel) [reply src/api/app/models/project.rb:1127]: Yeah I have some commits not pushed that change/test this. I got sidetracked by some nasty bug in `BsRequest.auto_accept`...

File: src/api/app/models/staging/workflow.rb
Comment (coolo) [reply src/api/app/models/staging/workflow.rb:118]: not in update_managers_group though

Comment (hennevogel) [reply src/api/app/models/staging/workflow.rb:118]: indeed

File: src/api/app/models/bs_request.rb
Comment (hennevogel) [inline src/api/app/models/bs_request.rb:512]: There are possibly more:

- BsRequest::Errors::CreatorCannotAcceptOwnRequests
- PostRequestMissingParameter

Not sure if we should catch them, sounds confusing to me...

Comment (hennevogel) [reply src/api/app/models/bs_request.rb:512]: or we catch in the policy directly...

File: src/api/app/policies/project_policy.rb
Comment (coolo) [inline src/api/app/policies/project_policy.rb:52]: note that force accept is no longer an option if you're not target maintainer

Comment (hennevogel) [reply src/api/app/policies/project_policy.rb:52]: As far as I can see this is just about https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request_permission_check.rb#L354-L355

Comment (coolo) [reply src/api/app/policies/project_policy.rb]: Jein. The policy uses force and skips the review check by that - but we later check that the project is in acceptable and so staging projects in review get blocked on a different level. The task of the policy is to check if it's technically possible, not if the staging project is "acceptable" or not.

But we also checked that the accepting user is target maintainer and so we *could* skip review checks. If this is no longer guaranteed, we need to check that the request is actually acceptable in full glory. I.e. if accepting the last review *really* will accept the request by approval.

Comment (hennevogel) [reply src/api/app/policies/project_policy.rb]: > The task of the policy is to check if it's technically possible, not if the staging project is "acceptable" or not.

The policy is about authorization, checking permissions. If someone is allowed to do something. But as you know things are tangled deeply...

>  if accepting the last review really will accept the request by approval

That is what we do no?

**Comment by hennevogel** [comment]
And now depends on #12845 12845

**Comment by hennevogel** [comment]
@coolo can you have another look please?

File: src/api/app/models/project.rb
Comment (hennevogel) [reply src/api/app/models/project.rb]: we talked offline

this is fine...

File: src/api/app/models/bs_request.rb
Comment (hennevogel) [reply src/api/app/models/bs_request.rb:501]: I decided against catching those "validations" in the permission check.

File: src/api/app/controllers/concerns/rescue_authorization_handler.rb
Comment (dmarcoux) [inline src/api/app/controllers/concerns/rescue_authorization_handler.rb:47]: A request state change means someone is potentially declining a request, not only accepting it. This message is confusing. How about this?

```suggestion
        "You cannot accept nor decline the request #{exception.record.number}."
```

**Review by coolo — APPROVED** [review]

File: src/api/app/controllers/concerns/rescue_authorization_handler.rb
Comment (hennevogel) [reply src/api/app/controllers/concerns/rescue_authorization_handler.rb:47]: This only ever happens in accepting staging projects

Comment (hennevogel) [reply src/api/app/controllers/concerns/rescue_authorization_handler.rb:47]: I can rename the `.reason` to be more staging related if you want to

**Comment by hennevogel** [comment]
Let's go then @adrianschroeter @coolo right?

**Comment by adrianschroeter** [comment]
sure, it works in production since quite some time.

**Comment by coolo** [comment]
This broke staging accept as the release tools always accept with force after checking on client side that everything is in acceptable. As acceptable is not in `FORCEABLE_STATES` this fails now. 

The reason we have to accept with force is that if you accept two stagings at the same time, the 2nd one will flip to 'building' due to the scheduler thinking "oh, something in Factory changed, let me check if all the projects linking to it need to change".

**Comment by DimStar77** [comment]
[That's already monkey patched on OBS
](https://github.com/openSUSE/open-build-service/pull/13004)

**Pull request merged**
